### ISSUE TITLE
`torch._C._scatter` becomes bottleneck for DataParallel

### ISSUE THREAD

**Issue by manuyavuz** [pytorch/pytorch#17209]
## 🐛 Bug

DataParallel runs slower on multi-GPU scenario because of slow scatter function.

## To Reproduce

1. Download PyTorch's DataParallel tutorial code: https://pytorch.org/tutorials/_downloads/39aeec3cba2dd8363e78683704cabea7/data_parallel_tutorial.py
1. run `CUDA_VISIBLE_DEVICES=0 python -m cProfile -s tottime data_parallel_tutorial.py > result_1` 
1. run `CUDA_VISIBLE_DEVICES=0,1 python -m cProfile -s tottime data_parallel_tutorial.py > result_2`
1. run `CUDA_VISIBLE_DEVICES=0,1,2,3 python -m cProfile -s tottime data_parallel_tutorial.py > result_4`
1. Check time difference.

Here are the most time-consuming operations:

```
##### 1 GPU #####
     224917 function calls (220822 primitive calls) in 3.474 seconds

  Ordered by: internal time

  ncalls  tottime  percall  cumtime  percall filename:lineno(function)
       6    2.683    0.447    2.687    0.448 {method 'to' of 'torch._C._TensorBase' objects}
     318    0.220    0.001    0.225    0.001 <frozen importlib._bootstrap_external>:830(get_data)
   29/28    0.166    0.006    0.169    0.006 {built-in method _imp.create_dynamic}
     318    0.038    0.000    0.038    0.000 {built-in method marshal.loads}
 878/875    0.031    0.000    0.047    0.000 {built-in method builtins.__build_class__}
    1386    0.030    0.000    0.030    0.000 {built-in method posix.stat}
      37    0.027    0.001    0.027    0.001 {built-in method posix.listdir}
       6    0.023    0.004    0.023    0.004 {built-in method torch._C._cuda_isDriverSufficient}

##### 2 GPUs ##### 
   226626 function calls (222502 primitive calls) in 6.569 seconds

 Ordered by: internal time

 ncalls  tottime  percall  cumtime  percall filename:lineno(function)
      4    2.764    0.691    2.764    0.691 {built-in method torch._C._scatter}
      6    2.645    0.441    2.645    0.441 {method 'to' of 'torch._C._TensorBase' objects}
   1386    0.225    0.000    0.225    0.000 {built-in method posix.stat}
    318    0.223    0.001    0.228    0.001 <frozen importlib._bootstrap_external>:830(get_data)
  29/28    0.160    0.006    0.166    0.006 {built-in method _imp.create_dynamic}
      8    0.063    0.008    0.063    0.008 {built-in method torch._C._broadcast_coalesced}
    318    0.048    0.000    0.048    0.000 {built-in method marshal.loads}
878/875    0.028    0.000    0.051    0.000 {built-in method builtins.__build_class__}
     13    0.024    0.002    0.024    0.002 {built-in method torch._C._cuda_isDriverSufficient}

##### 4 GPUs #####
   227666 function calls (223538 primitive calls) in 12.735 seconds

 Ordered by: internal time

 ncalls  tottime  percall  cumtime  percall filename:lineno(function)
      4    9.353    2.338    9.353    2.338 {built-in method torch._C._scatter}
      6    2.645    0.441    2.645    0.441 {method 'to' of 'torch._C._TensorBase' objects}
  29/28    0.157    0.005    0.162    0.006 {built-in method _imp.create_dynamic}
      8    0.111    0.014    0.111    0.014 {built-in method torch._C._broadcast_coalesced}
    318    0.063    0.000    0.067    0.000 <frozen importlib._bootstrap_external>:830(get_data)
    318    0.038    0.000    0.038    0.000 {built-in method marshal.loads}
   1386    0.028    0.000    0.028    0.000 {built-in method posix.stat}
     19    0.023    0.001    0.023    0.001 {built-in method torch._C._cuda_isDriverSufficient}
```

## Expected behavior

Multi-GPU run to be faster than single-GPU run.

## Environment

Please copy and paste the output from our
[environment collection script](https://raw.githubusercontent.com/pytorch/pytorch/master/torch/utils/collect_env.py)
(or fill out the checklist below manually).

You can get the script and run it with:
```
wget https://raw.githubusercontent.com/pytorch/pytorch/master/torch/utils/collect_env.py
# For security purposes, please check the contents of collect_env.py before running it.
python collect_env.py
```
```
Collecting environment information...



PyTorch version: 1.0.1.post2
Is debug build: No
CUDA used to build PyTorch: 9.0.176

OS: Ubuntu 16.04.5 LTS
GCC version: (Ubuntu 5.4.0-6ubuntu1~16.04.11) 5.4.0 20160609
CMake version: version 3.13.0

Python version: 3.6
Is CUDA available: Yes
CUDA runtime version: 7.5.17
GPU models and configuration:
GPU 0: Tesla K80
GPU 1: Tesla K80
GPU 2: Tesla K80
GPU 3: Tesla K80

Nvidia driver version: 410.79
cuDNN version: Probably one of the following:
/usr/lib/x86_64-linux-gnu/libcudnn.so.6.0.21
/usr/lib/x86_64-linux-gnu/libcudnn.so.7.0.5

Versions of relevant libraries:
[pip3] numpy==1.16.1
[pip3] torch==1.0.1.post2
[pip3] torchvision==0.2.1
[conda] pytorch-pretrained-bert   0.3.0                     <pip>
[conda] torch                     1.0.1.post2               <pip>
```

 - PyTorch Version (e.g., 1.0):
 - OS (e.g., Linux):
 - How you installed PyTorch (`conda`, `pip`, source):
 - Build command you used (if compiling from source):
 - Python version:
 - CUDA/cuDNN version:
 - GPU models and configuration:
 - Any other relevant information:

## Additional context

Might also be related to #14461

**Comment by manuyavuz** [comment]
Hi @mrshenli,

Do you have any guess on what could be the problem? Or any workaround that you might have?

I'm willing to help for any additional analysis/debug you might need. It's been a problem causing real slowness to us, We can't utilize our GPUs efficiently enough.

Also, we experience the same issue with a 2 V-100 GPU setting, so not sure if it's directly related to GPU type or latency that we might have.

**Comment by mrshenli** [comment]
@manuyavuz Sorry about the delay. I haven't got a chance to look into this yet. Will investigate later today.

**Comment by manuyavuz** [comment]
Thanks! Please tell if you would need any additional information.

**Comment by mrshenli** [comment]
Hi @manuyavuz, if I read it correctly, the tutorial is using a tiny model:

```python
input_size = 5
output_size = 2
self.fc = nn.Linear(input_size, output_size)
```

But it seems it took `2.7` seconds to copy the model from CPU to one GPU in your case. Do you consistently see that? Or are there any other workload running concurrently?

```
##### 1 GPU #####
     224917 function calls (220822 primitive calls) in 3.474 seconds

  Ordered by: internal time

  ncalls  tottime  percall  cumtime  percall filename:lineno(function)
       6    2.683    0.447    2.687    0.448 {method 'to' of 'torch._C._TensorBase' objects}
```

On my server, this copy call takes around 10 us.

```ipython
In [3]: linear = nn.Linear(5, 2)

In [4]: %timeit linear.cuda()
The slowest run took 4.57 times longer than the fastest. This could mean that an intermediate result is being cached.
10.1 µs ± 7.95 µs per loop (mean ± std. dev. of 7 runs, 1 loop each)
```

**Comment by mrshenli** [comment]
Sorry, that `2.7` seconds `to` invocation should be the data `input = data.to(device)` instead of model. But, `2.7` still seems too long.

```python
In [7]: x = torch.rand(100, 5)

In [8]: %timeit x.cuda()
17.4 µs ± 94.5 ns per loop (mean ± std. dev. of 7 runs, 100000 loops each)
```

**Comment by mrshenli** [comment]
I added timing measures into the tutorial code, and found that the first forward pass took `1.8` seconds. Subsequent forward passes took roughly 1ms each. It seems to be some initialization overhead, as you run it over more data. Do you see similar behavior in your case? @manuyavuz

**Comment by mrshenli** [comment]
I think it is indeed initialization overhead. I added `torch.tensor(1).cuda(0).cuda(1)` before the `for` loop to make sure that both GPUs are initialized. Then, all four forward pass execution times fall to a few milliseconds. 

I run the same program with your command on two GPUs, and I got:

```
##### 2 GPUs #####
   Ordered by: internal time

   ncalls  tottime  percall  cumtime  percall filename:lineno(function)
        6    1.463    0.244    1.463    0.244 {method 'to' of 'torch._C._TensorBase' objects}
        2    1.437    0.718    1.437    0.718 {method 'cuda' of 'torch._C._TensorBase' objects}
       42    0.323    0.008    0.323    0.008 {method 'acquire' of '_thread.lock' objects}
    33/32    0.147    0.004    0.149    0.005 {built-in method _imp.create_dynamic}

```

**Comment by manuyavuz** [comment]
Hi @mrshenli,

I could observe the same behavior. After dummy data assignment to GPUs, scatter works like a charm.

Do you think this thing warming step should be sth performed internally before doing any real work?

**Comment by mrshenli** [comment]
@manuyavuz I was expecting that `torch.cuda.init()` would do all necessary initialization, but it turns out that's not the case:

```python
import torch
torch.cuda.init()
x = torch.rand(1000, 1000)
y = x.to(0) # got ~ 1s delay here
```

@soumith  is that delay due to `CUcontext` init? Shall we cover that init in `torch.cuda.init()` as well (maybe by adding a `device` arg)?

**Comment by soumith** [comment]
The delay is most definitely CUcontext init. cc: @ezyang who wrote the lazy init stuff, he might have an opinion on whether to add CUcontext init in `torch.cuda.init()`

**Comment by ezyang** [comment]
`torch.cuda.init()` initializes *our* internal data structures, but not CUDA itself; that initialization happens automatically when you try to actually use the device. @ngimel, short of a kernel call, is there a way of forcing the actual CUDA initialization to occur?

In any case, I'm not sure it even makes sense to "fix" this, because if you're benchmarking, you should be running a few warmup loops of your timing loop before doing the real timings. If you agree with this, go ahead and close the issue; otherwise, I think it probably would be acceptable to add a dummy initialization call to torch.cuda.init, and I'll accept a patch that does so. Please also update the documentation to indicate that this will occur.

**Comment by ukemamaster** [comment]
@manuyavuz Did you solve this issue? 
I am having a similar kind of problem with `{method 'cuda' of 'torch._C._TensorBase' objects}` being the bottleneck in my code. I am using DDP but it happens even without DDP (on single GPU). 
Could you please share your solution, if you solved it?

**Comment by korotaS** [comment]
@ukemamaster not sure if it is still a problem after 3 years but I had a similar problem: during profiling I saw that `{method 'to' of 'torch._C._TensorBase' objects}` took around 2/3 of all my training time and this problem appeared after slight changes to my code, which I had to do to reproduce my experiments. I added this lines in the start of my script:
```python
torch.backends.cudnn.deterministic = True
torch.backends.cudnn.benchmark = False
```
After I added them, the whole script became much slower, and after I removed them (thanks to [this guide](https://pytorch.org/tutorials/recipes/recipes/tuning_guide.html#gpu-specific-optimizations)) - it all came back to normal. 
So, one possible solution is to **remove** those lines if you have them.
